src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,36 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            impInlineInfo->inlineResult->NoteFatal(InlineObservation::CALLEE_DOES_NOT_RETURN);
+            return;
+        }
+
```

Comment:
> But the point is that X is postdominated by a throw does not imply X is not performance sensitive.

I had similar doubts myself back when I did this experiment but I just can't find a reasonable use case that could be impacted by this. That X needs to be small enough for the method to be an inline candidate and it has to contain a loop otherwise the performance X will be dwarfed by the throw. And then the loop performance needs to be better when inlined due to optimizations such as constant propagation, otherwise the only impact we could possible measure would be the delay of the loop start by the time taken by the call.

I'd love to see an example that shows that this is indeed a problem. The current implementation already fails to reject inlining in some cases where there's no reason to inline, I'm somewhat reluctant to impose additional limitations.

> for instance methods that are loop-free

I'll have to check but I'm not sure there's any loop information available during inlining. Probably the next best thing would be to just match a single BB of type BBJ_THROW, possibly preceded by a list of BBJ_ALWAYS BBs. To make it even more conservative we could match only BBJ_THROW containing a single statement.

> and relatively small

Wouldn't this be redundant as the inliner already takes code size into account?